[WIP] Online editor for python

justwudi opened the pull request:
# Change existing form to React

The following partials have been moved to React:
- `_form.html.slim` -> `ProgrammingQuestionForm.jsx`
   Simple form has been removed, currently it contains just a `<div>` with a spinner that serves as the mounting point for the React component.
   The import alert has also been moved into `_props.json.jbuilder`
- `_form_build_log.html.slim` -> `BuildLog.jsx`
- `_form_templates.html.slim` -> `TemplatePackageView.jsx`
- `_form_test_cases.html.slim` -> `TemplateTestCaseView.jsx`

# Add online editor for python test cases

Users will be able to select to either use the online editor or upload a zip package when creating the programming question. This will be stored in in the database under a new column `package_type`. When editing the programming question, users will not be able to switch between the two types.

The online editor can be found in `OnlineEditorPythonView.jsx`. This view will be displayed when the user selects `python` as the language. Other editor views can be added in the future for other languages.

On the backend side, `Course::Assessment::Question::Programming::PackageGenerationConcern` will take in the `params` and construct the zip package from the fields of the online editor. This concern will route to `Course::Assessment::Question::Programming::Python::PackageConcern` when the language is `python`.

In the generated zip package, the fields used to generate the package are stored in a hidden json file `.meta`. This `.meta` file is used for displaying the fields for editing as well as to check whether there are any changes to the package when updating, the package will only be re-generated when the fields are modified.

TODO: 
- [ ] Fix tests
- [ ] Allow switching from online editor to zip package when updating
- [ ] Allow users to select between online editor and uploading zip package when they change the programming question to another language
- [ ] Allow users to attach other files that are used in the autograder. One challenge here is that we do not know whether to re-generate the package, maybe can store the hash of the file in the `.meta` file and compare them.

This closes #1568.

allenwq commented:
The PR is a bit too big. Here I give some general comments first:

1. The front-end validation is not done yet, I can submit the form and break the UI in some browser.

2. The fetch API does not seem to work on browsers like Safari, http://caniuse.com/#feat=fetch

3. Our front-end code is moving to material design,  better use the components there for styling for consistency.   Think `react-select` can be replaced with [material chip](http://www.material-ui.com/#/components/chip) ?

fonglh commented on app/controllers/concerns/course/assessment/question/programming/python/package_concern.rb:
```diff
@@ -0,0 +1,128 @@
+# frozen_string_literal: true
```
Create a new service for creating programming packages. Put this logic and the supporting template files with the service.

allenwq replied:
Yes, call it package generation service or something, so that you can properly test the service.

allenwq replied:
Also, I think you should document the input format of the service and the output package hierarchy.

fonglh commented on app/controllers/concerns/course/assessment/question/programming/python/package_concern.rb:
```diff
@@ -0,0 +1,128 @@
+            test_fn = "    def test_#{test_type.to_s}_#{index}(self):\n"\
+                      "        self.meta['expression'] = '#{escaped_expression}'\n"\
+
+            if test[:expected].first == '\'' && test[:expected].last == '\'' ||
```
This assumes that quote marks can only appear at the start and end of the string. Is this the intention?

allenwq commented on app/controllers/concerns/course/assessment/question/programming/python/package_concern.rb:
```diff
@@ -0,0 +1,128 @@
+
+      if tests.blank?
+        zip.print "    def test_public_1(self):\n"\
+                    "        self.meta['expression'] = 'This testcase has no meaning.'\n"\
```
These strings need to be translated. I am not sure if we need this, I think the UI should support a mode for non-autograded programming questions ?

allenwq commented on app/helpers/course/assessment/question/programming_helper.rb:
```diff
@@ -31,18 +33,51 @@ def display_build_log?
-      t('course.assessment.question.programming.form.import_result.success')
+  def successful_import_alert(json: false)
+    cls = ['alert', 'alert-success']
+    msg = t('course.assessment.question.programming.form.import_result.success')
```
Call it `klass` and `message`, We don't use shorthand variable names...

allenwq commented on client/app/bundles/course/assessment/question/programming.jsx:
```diff
@@ -0,0 +1,39 @@
+  );
+};
+
+$.getJSON('', (data) => {
```
How long does it take to build the data ?  If it's not very long it would be better to put the data in the initial request and parse them in the frontend, that will make the user side feels  faster.

LessonPlan pages uses 2 requests because the lesson plan data are quite large and it slows down the initial request a lot.

allenwq commented on client/app/bundles/course/assessment/question/programming/components/ProgrammingQuestionForm.jsx:
```diff
@@ -0,0 +1,339 @@
+              onSubmit={this.onSubmit.bind(this)} ref={(form) => { this.form = form; }} >
+          <input type='hidden' name='authenticity_token' value={formData.get('auth_token')} />
+
+          { this.renderInputField('title', 'Title', false, 'text', question.get('title') || '') }
```
All these strings need to be translated.

kxmbrian commented on client/app/bundles/course/assessment/question/programming/components/ProgrammingQuestionForm.jsx:
```diff
@@ -0,0 +1,339 @@
+
+const propTypes = {
+  data: React.PropTypes.shape({
+    question: PropTypes.object.isRequired,
```
Better to validate as `PropTypes.instanceOf(Immutable.Map).isRequired`

kxmbrian commented on client/app/bundles/course/assessment/question/programming/components/ProgrammingQuestionForm.jsx:
```diff
@@ -0,0 +1,339 @@
+
+  componentWillReceiveProps(nextProps) {
+    this.isLoading = nextProps.data.isLoading;
+    this.isEvaluating = nextProps.data.isEvaluating;
```
Is this needed? Possible to do without the assignment and check `this.props.isLoading` and `this.props.isEvaluating` directly?

kxmbrian commented on client/app/bundles/course/assessment/question/programming/containers/ProgrammingQuestion.jsx:
```diff
@@ -0,0 +1,117 @@
+      return <div className="alert alert-warning">Please select a language.</div>;
+
+    default:
+      return <div className="alert alert-info">Not yet implemented :(</div>;
```
Some untranslated strings here

kxmbrian commented on client/app/bundles/course/assessment/question/programming/reducers/programmingQuestionReducer.jsx:
```diff
@@ -0,0 +1,194 @@
+
+    case actionTypes.TEMPLATE_TAB_UPDATE:
+      const { selected } = action;
+      return $$state.set('package_ui', $$state.get('package_ui').set('selected', selected));
```
Try `$$state.setIn`

kxmbrian commented on client/app/bundles/course/assessment/question/programming/components/ProgrammingQuestionForm.jsx:
```diff
@@ -0,0 +1,339 @@
+  },
+  editTestsOnlineButton: {
+    id: 'course.assessment.question.programming.programmingQuestionForm.editTestsOnlineButton',
+    defaultMessage: 'Edit Tests Online',
```
The text might not be fitting in the case where we create tests. Will "Use Test Case Editor" or just "Test Case Editor" be better?

kxmbrian commented on client/app/bundles/course/assessment/question/programming/components/OnlineEditorPythonView.jsx:
```diff
@@ -0,0 +1,231 @@
+          value={this.props.data.get(field)}
+          onChange={this.onCodeChange.bind(this, field)}
+          editorProps={{$blockScrolling: true}}
+          setOptions={{useSoftTabs: true}}
```
Can set this as readOnly when `isLoading` also.

kxmbrian commented:
1. We are not using the `$$` convention anymore and are removing all instance of them. It seems it's like an arbitrary decision by `react_on_rails` and not a general standard.
2. Probably better for most users if we switch to the online editor mode when a language that has one is selected.
3. Run `npm run lint` from the `client` folder and fix the lints. You can `npm run lint-fix` to auto-fix some of them.
4. Haven't looked into the details, but I can't seem to add a programming question to an autograded assessment. Do you have the same issue?
5. Not done yet, but we'll need to switch to [`redux-immutable`](https://github.com/gajus/redux-immutable) since we are using immutable.js.

kxmbrian commented on client/app/bundles/course/assessment/question/programming/store.jsx:
```diff
@@ -0,0 +1,25 @@
+  };
+
+  const reducer = combineReducers(reducers);
+  const composedStore = compose(
```
Don't think the const name is appropriate here since the store is not composed yet.

allenwq commented:
@justwudi For your references, as discussed offline:
- UI/Service need to support create non-autograded programming questions 
- Need to handle the upload of data files and display exiting data files
- Use `axios` instead of `fetch`, [REF](https://github.com/mzabriskie/axios/issues/314)
- Only aync load the package, the form should be rendered first
- Remove react-select and use Material Chip ( think this can be generalised so that other question types can also use )

justwudi commented:
# New Changes

## Add support for non-autograded programming question

- Add checkbox in online editor mode for `autograded`
- If unchecked, only submission template field will be shown.
- Backend will remove any package attachments, clear existing templates, and add in the new submission template.

## Add support for data files

- Frontend
    + Current Data Files
        * Display filename and size of file
        * Checkbox to mark for deletion, default unchecked
    + New Data Files
        * List of input fields
        * New input field appears when user selected a file in the last input field
        * All input fields except for the last one has a delete button
- Backend
    + Unzips the current package, retrieves the list of files to put in the new package (those that are not mark for deletion)
    + New files will override old files if there are filename conflicts
    + `.meta` stores a list of `{ filename, size, hash }`, the hash is there so that we can make a proper comparison between the new `.meta` and the old `.meta` to determine whether there is a need to re-evaluate the package.

## Multiselect

- Implemented `ChipInput`, which is basically borrowed from [material-ui-chip-input](https://github.com/TeamWertarbyte/material-ui-chip-input) with minor changes to make it less buggy.
- Kind of works, with a few minor issues:
    + After clicking on an option, the dropdown closes. So the user will have to click it again.
    + Tapping on `Enter` after typing will clear the textfield and not add anything because no option is selected by default. The user will have to either click the option or use the up/down arrow keys to select an option before clicking on `Enter`

## Front end validation

- Added validation for `maximum_grade`, `language_id`.
- If in zip upload mode, validate that a package is uploaded if no package exists.
- If in editor mode and autograded is checked, validate that there exists at least 1 test case and that all test cases have both `expression` and `expected` filled in.

allenwq commented:
@justwudi Thanks for the effort !   

Possible to split this big PR into maybe 3+ smaller ones ?  You can make them depend on each other or even self contained. It will be easier for us to review and I believe it will make the code be merged faster as well.  Also when you fix something, you are supposed to rebase to the previous commit and fix them there.
